docs/proposals/08-shoot-additional-container-runtimes.md

```diff
@@ -0,0 +1,139 @@
+        1. A label for each container runtime support will be added to every node that belongs to the worker pool. This should be done similar
+           to the way labels created today for each node, through kubelet execution parameters (_kubelet.flags: --node-labels). When creating the OperatingSystemConfig (original) for the worker each container runtime support should be mapped to a label on the node.
+           For Example:
+           label: container.runtime.kata-containers=true (shoot.spec.cloud.<IAAS>.worker.containerRuntimes.kata-container)
+           label: container.runtime.gvisor=true (shoot.spec.cloud.<IAAS>.worker.containerRuntimes.gvisor)
+        2. During the Shoot reconciliation (Similar steps to the Extensions today) Gardener will create new RuntimeContainer resource if a container runtime exist in at least one worker spec:
+            ```yaml
+            apiVersion: runtimecontainer.gardener.cloud/v1alpha1
+            kind: RuntimeContainer
+            metadata:
+              name: kata-containers-runtime-extention
+              namespace: shoot--foo--bar
+            spec:
+              type: kata-containers
+            ```   
+   
+            Gardener will wait that all RuntimeContainers extensions will be reconciled by the appropriate extensions controllers.
+        3. Each runtime extension controller will be responsible to reconcile it's RuntimeContainer resource type.
+           rc-kata-containers extension controller will reconcile RuntimeContainer resource from type kata-container and rc-gvisor will reconcile RuntimeContainer resource from gvisor.
+           Reconciliation process by container runtime extension controllers:
+           1. Runtime extension controller from specific type should apply a chart which responsible for the installation of the runtime container in the cluster:
+                1. DaemonSet which will run a privileged pod on each node with the label: container.runtime.<type of the resource>:true The pod will be responsible for:
+                    1. Download the runtime container binaries (From the web or from the extension package   __TBD__) and copy them to the relevant path in the host OS.
+                    2. Add the relevant container runtime plugin section to the containerd configuration file (/etc/containerd/config.toml).
+                    3. Restart containerd in the node. 
+                3. RuntimeClass in the cluster to support the runtime class. for example:
+                   ```yaml
+                   apiVersion: node.k8s.io/v1beta1
+                   kind: RuntimeClass
+                   metadata:
+                     name: gvisor
+                   handler: runsc
+                   ```
+           2. Update the status of the relevant RuntimeContainer resource to succeeded.
+
+-->
+
+## Alternatives
```

Comment:
As written above, I am not convinced, we can abstract OS from container runtime. So, maybe we shouldn't try, because the proposal I see here just highlights the difficulties in doing so. Only when we have the container runtime, we can add stuff independent of the OS, but before that, the container runtime seems too strongly tied to the OS to be put away into a separate extension that requires awkward contracts to fuse them again together.

I don't want to bomb the idea, but please consider whether you propose an extension just because it's the norm/common or because it's actually independent enough to be an extension in the first place. If it's the latter, all is good. If it's the former, maybe we should merge the two and let the operating system/container runtime controller take care of both and advertise it's supported container runtimes instead.

Reply:
Please distinguish between two different things:

1. Container Runtime - Docker Engine/ContainerD/CRI-O
2. Runtime SHIM - gVisor/Kata Container/Docker

The first is tightly coupled to the OS. Our proposal starts by moving Kubernetes to ContainerD, as opposed to the current Docker Engine APIs. ContainerD is part of the Docker distribution, so the binaries are already in-place, and it's simply a matter of configuration.

Just solving the first item doesn't automatically mean that the actual containerization technique has changed. By default - if we move to containerd and do nothing else, we don't gain any additional runtime-classes.

By moving to a CRI implementation, we make it possible to configure the CRI implementation with additional runtime shims. These shims are the glue between the container management system and the additional runtimes we actually want - and they do not need to be tied to a specific OS.

Reply:
Thanks for the clarification @liorokman